### ISSUE TITLE
Sage Attention for diffuser library

### ISSUE THREAD

**Issue by ukaprch** [huggingface/diffusers#11168]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. I'm always frustrated when [...].

**Describe the solution you'd like.**

**Describe alternatives you've considered.**

**Additional context.**

**Comment by a-r-r-o-w** [comment]
Hey, adding different attention backends is going to be prioritized in the next release schedule! We're currently in the process of benchmarking and testing with different models to understand quality tradeoffs and performance gains.

For this specific case, using sage attention is as simple as replacing the F.scaled_dot_product_attention operation, and it should hopefully work for most, if not all, models out of the box

**Comment by ukaprch** [comment]
Sage attention on my pc needs to be disabled after the transformer does it's work when using the diffusers library. That was my experience at least.

**Comment by a-r-r-o-w** [comment]
Why exactly does it need to be disabled after transformer? Do you run into some error?

**Comment by ukaprch** [comment]
I ran into an error running the VAE afterwards. It (Sage Attn) complained about the dtype which is torch.float32. Apparently Sage Attn was still controlling something that affected the VAE. That was my experience. Why no one else may have this problem I can't tell you.

**Comment by ukaprch** [comment]
I ran a benchmark using sage attn on WAN 2.1 T2V 14B diffuser pipeline. I made some minor adjustments to generate 1 image (i.e. text to image vs text to video) to compare the usage of sage attention for this model. I was somewhat disappointed by the results leading me to believe that some of the claims being made out there are bupkus. I ran 5 test runs throwing out the 1st one as I did for the Flux 1.D model. I only noticed a 4.2% speed increase using sage attn vs not. My environment for Flux 1.D and WAN 2.1 are virtually the same. I have a quantized (qint8) transformer and T5TextEncoder for both setups.
Note: I am running with 64GB of RAM and an RTX 4090 with 24GB of VRAM. I am using Windows 10, python version 3.12.5 with CUDNN v 9.7 and torch-2.6.0+cu126.dist-info.

**Comment by bghira** [comment]
replacing the attn function isn't drop in. they take different parameters, and sageattn doesn't take attn mask or support dropout_p

**Comment by nitinmukesh** [comment]
Is Sage attention supported?
I checked but no information available
https://github.com/huggingface/diffusers/tree/main/docs/source/en/optimization

**Comment by a-r-r-o-w** [comment]
Sage attention support is being added in #11368

**Comment by a-r-r-o-w** [comment]
@varadrane1707 The PR is close to merge and only requires some discussion from other team members. For now, if you really need it, would recommend to create your own fork from the branch and modify it for your usecases. It is actually just a single file you can use with any code base, so feel free to copy-paste it and set `torch.nn.functional.scaled_dot_product_attention = dispatch_attention_fn`

**Comment by varadrane1707** [comment]
Thanks @a-r-r-o-w will try this out. Also did you check its compatibility with Para-attn since its used for fbcache implementation. I tried to use to use original sage attention implementation but the output video quality was bad for i2v 720p Wan2.1 model

**Comment by varadrane1707** [comment]
Hi , Tried with the branch , i am getting two issues 
One was the output video quality but i am looking that you are calling the vae decode outside the main pipe . Is it because of the attention backend changes? I am getting OOM because i am using the parallel vae but cant initialize parallel vae outside my main pipeline class. Any solution for that?

**Comment by a-r-r-o-w** [comment]
Hey @varadrane1707, it's hard for me to look at and understand any specific custom code you might be using. The best I can do is point you to a minified example that demonstrates how to use the attention dispatcher with context parallel (essentially the core of ParaAttention): https://gist.github.com/a-r-r-o-w/93b467ddf64bfe9df47fc12fc2ae4fac ; maybe you find it helpful

We have native support for FBC coming in #11180, which works without issues in my tests with the dispatcher, so that should be covered natively soon too.

re: sage attention output being bad -- if you're using custom code from different codebases, it will be hard for me to help. With diffusers-specific code, I haven't noticed a problem in the outputs. If you could open an issue with a simple code example, I could try to help.
